src/cmds/pbs_rsub.c

```diff
@@ -696,8 +765,158 @@ main(int argc, char *argv[], char *envp[])
 		}
 	}
 
+	if (is_maintenance_resv) {
+		int i;
+
+		pbs_errno = 0;
+		bstat_head = pbs_stathost(connect, "", NULL, NULL);
+		if (bstat_head == NULL) {
+			if (pbs_errno) {
+				errmsg = pbs_geterrmsg(connect);
+				if (errmsg != NULL) {
+					fprintf(stderr, "pbs_rsub: %s\n", errmsg);
+				} else {
+					fprintf(stderr, "pbs_rsub: Error (%d) submitting reservation\n", pbs_errno);
+				}
+			} else {
+				fprintf(stderr, "pbs_rsub: No nodes found\n");
+			}
+
+			CS_close_app();
+			exit(pbs_errno);
+		}
+
+		/* here, the select is crafted from hosts */
+		hostp = maintenance_hosts;
+		for (;*hostp ;hostp++) {
+			/* lets craft select_str */
+			for (bstat = bstat_head; bstat; bstat = bstat->next) {
+				for (pattr = bstat->attribs; pattr; pattr = pattr->next) {
+				    if (pattr->resource && strcmp(pattr->name, "resources_available") == 0 && strcmp(pattr->resource, "ncpus") == 0)
+					ncpus = pattr->value;
+				}
+
+				if (strcmp(bstat->name, *hostp) == 0 && atoi(ncpus) > 0) {
+					if (!select_str) {
+						/* +20 stands for "select=host=" + ":ncpus=" + '\0' */
+						select_str = (char*) malloc(strlen(bstat->name) + strlen(ncpus) + 20);
+						if (select_str == NULL) {
+							fprintf(stderr, "pbs_rsub: Out of memory\n");
+							exit(2);
+						}
+						sprintf(select_str, "select=host=%s:ncpus=%s", bstat->name, ncpus);
+					} else {
+						tmp_str = strdup(select_str);
+						if (tmp_str == NULL) {
+							fprintf(stderr, "pbs_rsub: Out of memory\n");
+							exit(2);
+						}
+						/* +14 stands for "+host=" + ":ncpus=" + '\0' */
+						select_str = (char*) realloc(select_str, strlen(tmp_str) + strlen(bstat->name) + strlen(ncpus) + 14);
+						if (select_str == NULL) {
+							fprintf(stderr, "pbs_rsub: Out of memory\n");
+							exit(2);
+						}
+						sprintf(select_str, "%s+host=%s:ncpus=%s", tmp_str, bstat->name, ncpus);
+						free(tmp_str);
+					}
+				}
+			}
+		}
+
+		pbs_statfree(bstat_head);	/* free info returned by pbs_stathost() */
+
+		pbs_errno = 0;
+		bstat_head = pbs_statvnode(connect, "", NULL, NULL);
```

Comment:
instead of using pbs_statvnode(), we can use pbs_stathost(). We can make the code simpler instead of re-inventing the wheel. What do you think?

Reply:
pbs_stathost() provides only list of hosts but we need the vnodes and its attributes because of crafting execvnodes. So, pbs_statvnode() is necessary.

I also use pbs_stathost() because we need to craft the select in the form like: "host=<host1>:ncpus=<ncpus_host1>...".

Maybe, we could use pbs_statvnode() and count the ncpus on all vnodes of the same host, but I think it would be more confusing than the used solution, and so it is better to call both pbs_stathost() and pbs_statvnode(). 

@prakashcv13 Do you think I should use only pbs_statvnode() and count the ncpus of the requested host from pbs_statvnode() (based on the same "host" attribute)?

Reply:
@vchlum - I should have spent more time in understanding why we are calling pbs_statvnode(). If we can avoid call to the pbs_stathost() and make the code better, we should do it. Could you please let me know why do you think using only pbs_statvnode() would be more confusing.

Reply:
@prakashcv13 I thought this way there will be two similar blocks of code with easier reading, but I thought about that and you are right. I will remove the calling of pbs_stathost(). It is not a problem. We should avoid calling server if possible.